src/dune_rules/obj_dir.ml

```diff
@@ -392,13 +392,21 @@ module Module = struct
     type t =
       | Immediate of Module.File.t
       | Transitive of Module.t * Ml_kind.t
+      | M2l of Module.t * Ml_kind.t
+      | Immediate_approx of Module.File.t
+      | Sig of Module.t
 
     let basename = function
       | Immediate f -> Path.basename (Module.File.path f) ^ ".d"
       | Transitive (m, ml_kind) ->
         let ext = sprintf ".%s.all-deps" (Ml_kind.to_string ml_kind) in
         let obj = Module.obj_name m in
         Module_name.Unique.artifact_filename obj ~ext
+      | M2l (m, ml_kind) ->
+        let ext = Ml_kind.choose ~impl:".m2l" ~intf:".m2li" ml_kind in
+        Module_name.to_string (Module.name m) ^ ext
+      | Immediate_approx f -> Path.basename (Module.File.path f) ^ ".d-approx"
+      | Sig m -> Module_name.to_string (Module.name m) ^ ".sig"
```

Comment:
I'm afraid something could break here for `M2l` and `Sig` by using the module name instead of its obj name (like `Transitive` does).
However, the obj names, e.g. `example__A` in `example__A.m2l` of a wrapped library, don't work for codept because it considers everything before the extension the module name.

Reply:
Indeed using `Module_name.to_string` is going to break in the presence of `(include_subdirs unqualified)`. Worth addinga a test for this.

@Octachron halp?

Reply:
Why would unqualified subdirs break it though? The same module name cannot occur in multiple subdirectories still.

I guess a workaround could be to output the m2ls/sigs all into directories based on the object name, but keeping the module name still in the basename.

Reply:
My bad, I was thinking of `(include_subdirs qualified)`. I'm actually working on this feature right now so it's on my mind.

Tbh, I'd still prefer if we used the object name as the prefix as it's consistent with all the other artifacts.

Reply:
For `sig` files, the name doesn't matter: signature files describe extension of the compilation unit environment and are thus self-contained.

For `.m2l*` files, it would be possible to record in the file the path of the compilation unit if it helps.
However, if dune can retrieve the path of the compilation unit in the later stage, one solution right now would be to store the file with dune preferred name and precise the compilation unit path of the file with the `file:path` syntax of codept when necessary.

Reply:
Thanks for the hint about `file:path`, that indeed works around the problem.

There still is one discrepancy though: the output of `-modules` still uses module names from filenames (recorded in `sig`s I guess), so the output is like `src/.example.objs/example__A.m2l: Example__B` even though `example__B.sig` is generated from `example__B.m2l:B`.

Maybe that is the right thing after all, if this is to generalize to `(include_subdirs qualified)`, but currently that's inconvenient to parse in dune, because there's no `Modules.find_dep` analogue for looking up modules by their object name. I currently implemented something very hacky to do that, just to confirm that it works.

Reply:
Both issues are fixed in `codept` master branch. I would be kind of surprised that those are the only issues, so I am planning to release the `0.12` version once all remaining issues with dune integration are ironed out.

Reply:
Thanks! That simplifies things here indeed.

I now merged dune's `main` into this PR and resolved the conflicts due to `(include_subdirs qualified)`. Looks like the tests for qualified still pass with codept, so that's good. Although the namespace and nesting power of codept isn't yet being taken advantage of for those.
Not sure if I should try to figure that out here or leave it as follow-up like other aspects related to qualified were.

`-inner-modules` would make sense and I tried it, but it doesn't fully work right now with virtual libraries. Some extra logic is still required to deal with `.sig`s between virtual libraries and their implementations.